File: providers/aws/src/test/java/cloudspec/aws/ec2/EC2RouteTableLoaderTest.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2;

import cloudspec.aws.ec2.loader.EC2RouteTableLoader;
import datagen.services.ec2.model.RouteTableGenerator;
import org.junit.Test;
import software.amazon.awssdk.services.ec2.model.DescribeRouteTablesRequest;
import software.amazon.awssdk.services.ec2.model.DescribeRouteTablesResponse;
import software.amazon.awssdk.services.ec2.model.RouteTable;

import java.util.function.Consumer;

import static org.junit.Assert.*;
import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doAnswer;

public class EC2RouteTableLoaderTest extends EC2LoaderTest {
    private final EC2RouteTableLoader loader = new EC2RouteTableLoader(clientsProvider);

    @Test
    public void shouldLoadRouteTables() {
        var routeTables = RouteTableGenerator.routeTables();
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeRouteTablesRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeRouteTablesRequest.builder();
            builderLambda.accept(builder);
            assertTrue(builder.build().filters().isEmpty());
            return DescribeRouteTablesResponse.builder()
                                              .routeTables(routeTables.toArray(new RouteTable[0]))
                                              .build();
        }).when(ec2Client).describeRouteTables(any(Consumer.class));

        var resources = loader.getAll();

        assertNotNull(resources);
        assertEquals(routeTables.size(), resources.size());
        assertTrue(routeTables.containsAll(resources));
    }

    @Test
    public void shouldNotLoadRouteTableByRandomId() {
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeRouteTablesRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeRouteTablesRequest.builder();
            builderLambda.accept(builder);
            assertFalse(builder.build().filters().isEmpty());
            return DescribeRouteTablesResponse.builder()
                                              .build();
        }).when(ec2Client).describeRouteTables(any(Consumer.class));

        var resourceOpt = loader.getById(RouteTableGenerator.routeTableId());

        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isEmpty());
    }

    @Test
    public void shouldLoadRouteTableById() {
        var routeTables = RouteTableGenerator.routeTables(1);
        doAnswer(answer -> {
            var builderLambda = (Consumer<DescribeRouteTablesRequest.Builder>) answer.getArguments()[0];
            var builder = DescribeRouteTablesRequest.builder();
            builderLambda.accept(builder);
            assertFalse(builder.build().filters().isEmpty());
            return DescribeRouteTablesResponse.builder()
                                              .routeTables(routeTables.toArray(new RouteTable[0]))
                                              .build();
        }).when(ec2Client).describeRouteTables(any(Consumer.class));

        var routeTable = routeTables.get(0);
        var resourceOpt = loader.getById(routeTable.routeTableId());
        assertNotNull(resourceOpt);
        assertTrue(resourceOpt.isPresent());
        assertEquals(resourceOpt.get(), routeTable);
    }
}

File: providers/aws/src/test/kotlin/cloudspec/aws/ec2/EC2RouteTableLoaderTest.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2

import arrow.core.None
import arrow.core.Some
import datagen.services.ec2.model.RouteTableGenerator.routeTableId
import datagen.services.ec2.model.RouteTableGenerator.routeTables
import io.mockk.every
import org.junit.Test
import software.amazon.awssdk.services.ec2.model.DescribeRouteTablesRequest
import software.amazon.awssdk.services.ec2.model.DescribeRouteTablesResponse
import java.util.function.Consumer
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertTrue

class EC2RouteTableLoaderTest : EC2LoaderTest() {
    private val loader = EC2RouteTableLoader(clientsProvider)

    @Test
    fun `should load all resources`() {
        val routeTables = routeTables()

        every {
            ec2Client.describeRouteTables(any<Consumer<DescribeRouteTablesRequest.Builder>>())
        } answers {
            val builder = DescribeRouteTablesRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeRouteTablesRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertTrue(request.filters().isEmpty())

            DescribeRouteTablesResponse.builder()
                    .routeTables(routeTables)
                    .build()
        }

        val resources = loader.all
        assertNotNull(resources)
        assertEquals(routeTables.size, resources.size)
    }

    @Test
    fun `should not load resource by random id`() {
        every {
            ec2Client.describeRouteTables(any<Consumer<DescribeRouteTablesRequest.Builder>>())
        } answers {
            val builder = DescribeRouteTablesRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeRouteTablesRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeRouteTablesResponse.builder()
                    .build()
        }

        val resourceOpt = loader.byId(routeTableId())
        assertTrue(resourceOpt is None)
    }

    @Test
    fun `should load resource by id`() {
        val routeTables = routeTables()

        every {
            ec2Client.describeRouteTables(any<Consumer<DescribeRouteTablesRequest.Builder>>())
        } answers {
            val builder = DescribeRouteTablesRequest.builder()

            val builderLambda = args[0] as Consumer<DescribeRouteTablesRequest.Builder>
            builderLambda.accept(builder)

            val request = builder.build()
            assertFalse(request.filters().isEmpty())

            DescribeRouteTablesResponse.builder()
                    .routeTables(routeTables)
                    .build()
        }

        val routeTable = routeTables[0]
        val resourceOpt = loader.byId(routeTable.routeTableId())
        assertTrue(resourceOpt is Some<*>)
    }
}
